File: plc4j/api/src/main/java/org/apache/plc4x/java/api/connection/PlcConnection.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/
package org.apache.plc4x.java.api.connection;

import org.apache.plc4x.java.api.exceptions.PlcConnectionException;
import org.apache.plc4x.java.api.exceptions.PlcException;
import org.apache.plc4x.java.api.model.Address;

import java.util.Optional;

/**
 * Interface defining the most basic methods a PLC4X connection should support.
 * This generally handles the connection establishment itself and the parsing of
 * address strings to the platform dependent Address instances.
 *
 * The individual operations are then defined by other interfaces within this package.
 */
public interface PlcConnection extends AutoCloseable {

    /**
     * Established the connection to the remote PLC.
     *
     * @throws PlcConnectionException an exception if the connection attempt failed.
     */
    void connect() throws PlcConnectionException;

    /**
     * Closes the connection to the remote PLC.
     *
     * @throws Exception an exception if shutting down the connection failed.
     */
    @Override
    void close() throws Exception;

    /**
     * Parses a PLC/protocol dependent address string into an Address object.
     *
     * @param addressString String representation of an address for the current type of PLC/protocol.
     * @return Address object identifying an address for the current type of PLC/protocol.
     * @throws PlcException an exception if there was a problem parsing the address string.
     */
    Address parseAddress(String addressString) throws PlcException;

    Optional<PlcLister> getLister();

    Optional<PlcReader> getReader();

    Optional<PlcWriter> getWriter();

}

File: sandbox/plc4cpp/api/src/main/cpp/org/apache/plc4x/cpp/PlcConnection.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
 
#include "PlcConnection.h"
#include "Logger.h"

org.apache.plc4x.java.api

namespace org
{
  namespace apache
  {
    namespace plc4x
    {
	    namespace cpp
	    {
			namespace api
			{

				PlcConnection::PlcConnection() :
					_bOpen(false),
					_pSendBuffer(NULL),
					_pRecvBuffer(NULL),
					_iSendBufferSize(0),
					_iRecvBufferSize(0),
				{
					setSendBufferSize(DEFAULT_BUFFER_SIZE);
					setRecvBufferSize(DEFAULT_BUFFER_SIZE);
					_strErrorMessage = "";
				}

				PlcConnection::~PlcConnection()
				{
					if (_pSendBuffer != NULL) delete[] _pSendBuffer;
					if (_pRecvBuffer != NULL) delete[] _pRecvBuffer;
				}

				PlcField PlcConnection::prepareField(String fieldQuery) throws PlcInvalidFieldException 
				{
						BOOST_THROW_EXCEPTION(PlcRuntimeException("Parse method is not implemented for this connection / driver"));
				}
				
				void PlcConnection::setSendBufferSize(int iBufferSize)
				{
					if (iBlockSize != _iSendBufferSize)
					{
						_iSendBufferSize = iBlockSize;
						if (_pSendBuffer != NULL) delete[] _pSendBuffer;
						_pSendBuffer = new unsigned char[_iBlockSizeSend];
					}
				}

				void PlcConnection::setRecvBufferSize(int iBufferSize)
				{
					if (iBlockSize != _iRecvBufferSize)
					{
						_iRecvBufferSize = iBufferSize;
						if (_pRecvBuffer != NULL) delete[] _pRecvBuffer;
						_pRecvBuffer = new unsigned char[_iBlockSizeReceive];
					}
				}
			}
	    }
    }
  }
}
